Automattic/woocommerce-payments
# Switch to proper input element when using Stripe Link autofill

Closes https://github.com/Automattic/woocommerce-payments/issues/7791

#### Changes proposed in this Pull Request
This PR
* ensures that with saved payment tokens on the shortcode checkout page, when autofilling payment details with Link, the input element selected is switched from the default saved payment token to the auto-filled Stripe Link element
* switches from the default saved payment token to new payment token when users click the Stripe Link button within the email field
* focuses on introducing the above two changes in the shortcode checkout, because Blocks checkout stays unaffected due to different nature of rolling out Stripe Link on the page (it does so only after users select new payment token)










#### Testing instructions

> please remove all saved payment methods before starting.




* open the shortcode checkout and provide a non-registered Stripe Link email: you should be able to see the _save to Link_ checkbox after providing the card details. Select this checkbox & save details to Link.
* prepare another cart and open the shortcode checkout: click the Link button in the email element and confirm that autofill works as expected. Complete the transaction by either closing the page or purchasing the product.
* save a new payment method either by navigating to _My account -> Payment methods_ or by saving payment method during the payment. Make sure that the saved payment method is set as default.
* go to checkout page, confirm the saved payment token is selected by default
    * click Stripe Link button and verify that together with Link modal popping up, the selected payment method is switched from saved token to the new payment element which is used for injecting details from Link after completing its autofill modal 
    * complete the autofill modal steps and confirm you have the right payment method selected
    * complete the purchase, navigate to order details in the admin panel, open payment intent details and confirm payment method is Link 
    * confirm payment method is Link on Stripe side
* navigate to shortcode checkout page again but this time just pay with the saved payment token (card) which is selected by default
    * in the order details -> payment intent details, confirm that the payment method is now CC
    * confirm payment method is CC on Stripe side
 * ensure that the non-registered Stripe Link email is used on the shortcode checkout page and open it
 * the saved card should be now still selected by default
 * click the Stripe Link button and confirm that the input element selected is switched from the default saved payment token to the new payment token
 * provide card details and confirm you now see the _save to Link_ checkbox
 * use Stripe Link in Blocks checkout and confirm that main flows from above are working correctly.
     * among others, check the following specific scenario: while having saved payment tokens, 
         * open Blocks checkout & confirm the Stripe Link button is not displayed unless you click the new card element
         * switch back to selecting the saved payment token and confirm the Link button was removed (more details in https://github.com/Automattic/woocommerce-payments/issues/7791#issuecomment-1892418841 and https://github.com/Automattic/woocommerce-payments/issues/7791#issuecomment-1893639593)
         * choose new card payment method again and confirm Stripe Link button is back again and can be used without problems


-------------------


**Post merge**
